### ISSUE TITLE
Fallback to site DB on searches by username

### ISSUE THREAD

**Issue by mbaruh** [python-discord/bot#1757]
We sometimes need to query user data such as the user or infractions embed, when all we have is the username.

Since the site DB stores user ID's and usernames, and already has an endpoint to retrieve stored data based on username, we can use that in case the user is not in the guild and we try to query by their username, to get the user's ID.

This can be implemented by improving the FetchedUser converter to handle cases where the given arguement doesn't represent an integer.

**Comment by MarkKoz** [comment]
@mbaruh The fetch API call already provides a `User` object with a `name` attribute filled in. Did you mean "nickname" rather than "username"?

**Comment by MarkKoz** [comment]
Also, can you provide an example of a situation in which you are only able to identify someone by their user/nickname? If it's coming from bot output, then appending the ID to those outputs or using mentions can be considered.

**Comment by Akarys42** [comment]
> Also, can you provide an example of a situation in which you are only able to identify someone by their user/nickname? If it's coming from bot output, then appending the ID to those outputs or using mentions can be considered.

Most of the time it is in a ModMail thread, where a user would report a user who left in the meantime for instance.

**Comment by mbaruh** [comment]
@MarkKoz the `UserConverter` relies on the cache, which is usually not enough. Whenever we try look someone up by name#discrim when they've left the guild we get an error.

**Comment by MarkKoz** [comment]
> 
> 
> @MarkKoz the `UserConverter` relies on the cache, which is usually not enough. Whenever we try look someone up by name#discrim when they've left the guild we get an error.

It doesn't anymore; see #1742

**Comment by mbaruh** [comment]
I see. If after merging it's resolved I'll close the issue then.

**Comment by TizzySaurus** [comment]
> @MarkKoz the `UserConverter` relies on the cache, which is usually not enough. Whenever we try look someone up by name#discrim when they've left the guild we get an error.

From my understanding this will still be the case. The `UserConverter` only doesn't rely on cache when the user id is passed. If you want name#discrim support for non-cached users then the `UserConverter` doesn't give this (again, if I'm correctly understanding everything).

**Comment by mbaruh** [comment]
From my testing the converters PR does not achieve the desired behavior described here.

**Comment by TizzySaurus** [comment]
How exactly are we planning on implementing this functionality?

My thoughts currently is that rather than adding another converter, we'd allow the passed user to be a `str` and in the case it is a string, we call some utility function which will query the database -- the utility function returning either `None` or raising an error when the username isn't found and returning the wanted data (e.g. infractions) when the username is found.

**Comment by MarkKoz** [comment]
Why should it not just be a converter? What advantage is there to invoking it manually?

**Comment by TizzySaurus** [comment]
Making it a converter brings other complications, such as what would the converter return?

From my understanding it'd have to end up returning `t.Union[MemberOrUser, discord.Object]` when the person is valid since we want to allow for deleted accounts but then this leads to a typehint being of two different types, potentially leading to further implications.

There's also the fact that we'd be querying the database twice (once for the conversion, and a potential second time for the actual query of command)

**Comment by mbaruh** [comment]
> 
> 
> Making it a converter brings other complications, like what would it return?
> 
> I suppose we'd want the converter to return a custom datatype instance which has all of the user data obtainable from the database?
> 
> I'm struggling to envisage how exactly this would work.

The way it would work is that it would try to find the username and discrim in the database, and if it's there fetch the ID. Once you have the ID you can use the discord API as usual.

**Comment by TizzySaurus** [comment]
Right, but what about when it's a deleted account? The user id will return but converting to a User object will fail (would have to have a fallback of `discord.Object`, meaning the converter would have two possible return types)

**Comment by mbaruh** [comment]
It should just fail to convert in that case.
